### PR TITLE
Timeline/Status grid panel: Add tooltip support

### PR DESCRIPTION
**Pull request by dprokop**
First version, ready to test. But some more work there to be done.

Timeline:
image

Status grid:
image

TODO:
- [x] Highlight doesn't work when tooltip is disabled (because highlight and tooltip positioning happen in the same setCursor hook which is no there when tooltip is disabled)
- [ ] Changing tooltip mode is not reflected WHEN IN EDIT MODE. This applies to all GraphNG based visualizations - will fix separately.

### CODE REVIEW

File: public/app/plugins/panel/state-timeline/timeline.ts [hunk 1]
```diff
@@ -382,7 +381,7 @@ export function getConfig(opts: TimelineCoreOptions) {
     hovered[i] = o;
   }
 
-  function hoverMulti(cx: number, cy: number) {
+  function highlightMulti(cx: number, cy: number) {
```
Comment (dprokop) [inline public/app/plugins/panel/state-timeline/timeline.ts:384]: @leeoniya I've done couple of naming changes here. I want to separate tooltip from highlight and this is the first step. Happy to hear some thoughts!

File: public/app/plugins/panel/status-history/module.tsx [hunk 2]
```diff
@@ -75,5 +74,6 @@ export const plugin = new PanelPlugin<StatusPanelOptions, StatusFieldConfig>(Sta
         },
       });
 
-    addLegendOptions(builder, false);
+    commonOptionsBuilder.addLegendOptions(builder, false);
+    commonOptionsBuilder.addTooltipOptions(builder, true);
```
Comment (dprokop) [inline public/app/plugins/panel/status-history/module.tsx:78]: Disabling "Multi" tooltip mode

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 3]
```diff
@@ -32,6 +41,69 @@ export const StateTimelinePanel: React.FC<TimelinePanelProps> = ({
...
+          <br />
+          <SeriesTableRow label={display.text} color={display.color || FALLBACK_COLOR} isActive />
+          From <strong>{xFieldFmt(xField.values.get(datapointIdx!)).text}</strong>
+          {toFragment}
+          {durationFragment}
+        </div>
+      );
+    },
+    [theme, timeZone, data]
+  );
```
Comment (dprokop) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:105]: Custom tooltip renderer for Timeline panel - can be extracted to a separate component

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 4]
```diff
@@ -1,9 +1,18 @@
...
+  getDisplayProcessor,
+  getFieldDisplayName,
+  PanelProps,
+} from '@grafana/data';
+import { TooltipDisplayMode, TooltipPlugin, useTheme2, ZoomPlugin } from '@grafana/ui';
 import { TimelineMode, TimelineOptions } from './types';
 import { TimelineChart } from './TimelineChart';
-import { prepareTimelineFields, prepareTimelineLegendItems } from './utils';
+import { findNextStateIndex, prepareTimelineFields, prepareTimelineLegendItems } from './utils';
+import { Interval, toDuration } from '@grafana/data/src/valueFormats/dateTimeFormatters';
```
Comment (torkelo) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:14]: you do not need to import this, you can import getValueFormat and then call getValueFormat('dtdurationms');   get the same formatter you are using below

Comment (dprokop) [reply public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:14]: Ah, awesome! Thanks @torkelo

**Review by torkelo — APPROVED** [review]
Awesome job! works great.

**Review by ryantxu — APPROVED** [review]
This is great!   lets merge for 8.0

File: public/app/plugins/panel/state-timeline/timeline.ts [hunk 5]
```diff
@@ -406,25 +403,22 @@ export function getConfig(opts: TimelineCoreOptions) {
...
 
-    if (found) {
-      setHoverMark(0, found);
-    } else if (hovered[0] != null) {
-      setHoverMark(0, null);
-    }
+    return found;
   }
 
-  const doHover = mode === TimelineMode.Changes ? hoverMulti : hoverOne;
```
Comment (leeoniya) [inline public/app/plugins/panel/state-timeline/timeline.ts:425]: this was here specifically to remove the mode comparison test during each mousemove event

File: public/app/plugins/panel/state-timeline/timeline.ts (hunk 1 above)
Comment (leeoniya) [reply public/app/plugins/panel/state-timeline/timeline.ts:382]: renames look mostly fine to me

**Review by leeoniya — APPROVED** [review]

File: public/app/plugins/panel/state-timeline/timeline.ts [hunk 6]
```diff
@@ -47,6 +47,8 @@ export interface TimelineCoreOptions {
   getTimeRange: () => TimeRange;
   formatValue?: (seriesIdx: number, value: any) => string;
   getFieldConfig: (seriesIdx: number) => TimelineFieldConfig;
+  onHover?: (seriesIdx: number, valueIdx: number, rect: Rect) => void;
```
Comment (ryantxu) [inline public/app/plugins/panel/state-timeline/timeline.ts:50]: should we make them required

Comment (leeoniya) [reply public/app/plugins/panel/state-timeline/timeline.ts:50]: :+1: , will do in followup

**Pull request merged**